### PR TITLE
allow changing number of threads through the algorithm settings panel

### PR DESCRIPTION
**Pull request by alexbruy**
# Description

Handy if algorithm is capable of using several cores (e.g. TauDEM, PDAL).

image

Also update PDAL algorithms and TileXYZ algorithm to use this new setting.

### CODE REVIEW

**Comment by nyalldawson** [comment]
I gather the setting is currently unused anywhere?

Can you move the widget up into the "environment settings" group? (Feel free to rename that group to something more generic if you want!)

**Comment by alexbruy** [comment]
> I gather the setting is currently unused anywhere?

It can be used by PDAL algorithms and native "Generate XYZ tiles" algorithms. I haven't updates algorithms, as I was thinking that keeping things isolated will make review simpler. Will be happy to add these changes to this PR if you prefer.

> Can you move the widget up into the "environment settings" group?

Done

**Comment by alexbruy** [comment]
@nyalldawson I have updated PR to use this setting in TileXYZ algorithm and in PDAL provider.

File: src/core/processing/qgsprocessingcontext.h [hunk 1]
```diff
@@ -680,6 +681,30 @@ class CORE_EXPORT QgsProcessingContext
...
+     * Returns the (optional) number of threads to use when running algorithms.
+     *
+     * If set, this overrides the standard global Processing number of threads setting.
+     * Note that if algorithm implementation does not support multhreaded execution, this
+     * setting will be ignored.
+     *
+     * \see setNumberOfThreads()
+     * \since QGIS 3.32
+     */
+    int numberOfThreads() const;
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingcontext.h:694]: maybe this should be "maximumThreads" instead? It's more reflective of the actual use of the setting.

File: src/core/processing/qgsprocessingcontext.h [hunk 2]
```diff
@@ -680,6 +681,30 @@ class CORE_EXPORT QgsProcessingContext
      */
     void setTemporaryFolder( const QString &folder );
 
+    /**
+     * Returns the (optional) number of threads to use when running algorithms.
+     *
+     * If set, this overrides the standard global Processing number of threads setting.
+     * Note that if algorithm implementation does not support multhreaded execution, this
+     * setting will be ignored.
+     *
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingcontext.h:690]:
```suggestion
     * \warning Not all algorithms which support multithreaded execution will respect this setting, depending on the multi-threading framework in use. Multithreaded algorithms must check this value and adapt their thread handling accordingly -- the setting will not be automatically applied.
```

**Comment by nyalldawson** [comment]
Looks fine to me. Just a word of caution that it could be tricky to correctly support this in some algorithms -- eg something like QgsRasterizeAlgorithm uses multithreading via QtConcurrent, which means that it's using the global Qt thread pool max threads setting. And that thread pool is shared with rendering and everything else QGIS does... so if we try to change the maximum threads for that algorithm, we'd end up also impacting a whole bunch of non-processing related thread stuff.

(I think to handle this we'd also need to add a separate QThreadPool to the context for use by processing only, and ensure this is used by multithreaded algorithms....)

File: src/core/processing/qgsprocessingcontext.h (hunk 2 above)
Comment (alexbruy) [reply src/core/processing/qgsprocessingcontext.h:691]: Added, thanks!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**